internal/datastore/mysql/migrations/driver.go

```diff
@@ -35,16 +37,29 @@ type MySQLDriver struct {
 //
 // URI: [scheme://][user[:[password]]@]host[:port][/schema][?attribute1=value1&attribute2=value2...
 // See https://dev.mysql.com/doc/refman/8.0/en/connecting-using-uri-or-key-value-pairs.html
-func NewMySQLDriverFromDSN(url string, tablePrefix string) (*MySQLDriver, error) {
+func NewMySQLDriverFromDSN(url string, tablePrefix string, credentialsProvider datastore.CredentialsProvider) (*MySQLDriver, error) {
 	dbConfig, err := sqlDriver.ParseDSN(url)
 	if err != nil {
 		return nil, fmt.Errorf(errUnableToInstantiate, err)
 	}
 
-	db, err := sql.Open("mysql", dbConfig.FormatDSN())
+	if credentialsProvider != nil {
+		// add a before connect callback to trigger the token retrieval from the credentials provider
+		err := dbConfig.Apply(sqlDriver.BeforeConnect(func(ctx context.Context, config *sqlDriver.Config) error {
+			config.User, config.Passwd, err = credentialsProvider.Get(ctx, config.Addr, config.User)
+			return err
+		}))
+		if err != nil {
+			return nil, fmt.Errorf(errUnableToInstantiate, err)
+		}
+	}
+
+	connector, err := sqlDriver.NewConnector(dbConfig)
```

Comment:
Similarly here, we create the db using `sqlDriver.NewConnector(dbConfig)` and `db := sql.OpenDB(connector)` instead of `sql.Open("mysql", dbConfig.FormatDSN())`, so we can preserve the `BeforeConnect` hook we added